sdk/core/core-amqp/review/core-amqp.api.md

```diff
@@ -491,7 +491,7 @@ export { isAmqpError }
 export function isIotHubConnectionString(connectionString: string): boolean;
 
 // @public
-export function isMessagingError(error: Error | MessagingError): error is MessagingError;
+export function isMessagingError<knownErrorCodesOnly extends true | false = true>(error: Error | MessagingError<knownErrorCodesOnly>): error is MessagingError<knownErrorCodesOnly>;
```

Comment:
@bterlson  - what do you think of this particular trick?

The constraint here is that `MessagingError` was already exported (and obviously, exported with 'code' being a plain string). 

I think we've basically got these cases:
1. Users using MessagingError prior to any of our code (isMessagingError didn't exist)
2. Users that are _now_ going to use MessagingError

For #1, I've left MessagingError (with a default type parameter) with a 'code' field that is of type string. This makes it compatible with anyone that had no type expectations.

For #2 I've made it so isMessagingError (which is newly added with this revision), by default, returns a MessagingError with 'code' typed _only_ with the string union types we've defined in MessagErrorCodes.

Does this usage seem okay? The problem here is that, technically, the .code is not fixed in place and could have additional values added to it. However, there are a core set that a user would be expected to potentially programatically react to so I'd like them to be front-and-center and easily accessible with tooling.

Reply:
This idea of using the custom type guard to narrow into a more precise type than existed previously is clever, but what about a simpler implementation: have `interface MessagingErrorWithKnownCodes extends MessagingError` that has the strict string literal enum, and have your type guard coerce into that type?

Reply:
I'm going to remove this for now but I'll take your idea into consideration when I take a run at this again.